contracts/shapella-upgrade/ShapellaUpgradeTemplate.sol

```diff
@@ -0,0 +1,374 @@
+        _transferOZAdminFromThisToVoting(_locator.stakingRouter());
+        _transferOZAdminFromThisToVoting(_locator.accountingOracle());
+        _transferOZAdminFromThisToVoting(_locator.validatorsExitBusOracle());
+        _transferOZAdminFromThisToVoting(_locator.withdrawalQueue());
+
+        IOssifiableProxy(_locator.stakingRouter()).proxy__changeAdmin(_voting);
+        IOssifiableProxy(_locator.accountingOracle()).proxy__changeAdmin(_voting);
+        IOssifiableProxy(_locator.validatorsExitBusOracle()).proxy__changeAdmin(_voting);
+        IOssifiableProxy(_locator.withdrawalQueue()).proxy__changeAdmin(_voting);
+    }
+
+    function _verifyUpgrade() internal view {
+        require(IVersioned(_locator.legacyOracle()).getContractVersion() == 4, "INVALID_LO_VERSION");
+        require(IVersioned(_locator.accountingOracle()).getContractVersion() == 1, "INVALID_AO_VERSION");
+        require(IVersioned(_locator.stakingRouter()).getContractVersion() == 1, "INVALID_SR_VERSION");
+        require(IVersioned(_locator.validatorsExitBusOracle()).getContractVersion() == 1, "INVALID_EB_VERSION");
+        require(IVersioned(_locator.withdrawalQueue()).getContractVersion() == 1, "INVALID_WQ_VERSION");
+
+        require(IOssifiableProxy(address(_locator)).proxy__getAdmin() == _voting, "INVALID_LOCATOR_ADMIN" );
+        require(IOssifiableProxy(_locator.accountingOracle()).proxy__getAdmin() == _voting, "INVALID_AO_PROXY_ADMIN");
+        require(IOssifiableProxy(_locator.stakingRouter()).proxy__getAdmin() == _voting, "INVALID_SR_PROXY_ADMIN");
+        require(IOssifiableProxy(_locator.validatorsExitBusOracle()).proxy__getAdmin() == _voting, "INVALID_EB_PROXY_ADMIN");
+        require(IOssifiableProxy(_locator.withdrawalQueue()).proxy__getAdmin() == _voting, "INVALID_WQ_PROXY_ADMIN");
+
+        require(IPausableUntil(_locator.validatorsExitBusOracle()).isPaused(), "EB_NOT_PAUSED");
+        require(IPausableUntil(_locator.withdrawalQueue()).isPaused(), "WQ_NOT_PAUSED");
+        require(IPausableUntil(_locator.withdrawalQueue()).getResumeSinceTimestamp() == _hardforkTimestamp, "INCORRECT_WQ_RESUME_SINCE_TIMESTAMP");
+    }
+
+    function _transferOZAdminFromThisToVoting(address _contract) internal {
+        bytes32 adminRole = IAccessControl(_contract).DEFAULT_ADMIN_ROLE();
+        IAccessControl(_contract).grantRole(adminRole, _voting);
+        IAccessControl(_contract).renounceRole(adminRole, address(this));
+    }
+
+    function _pauseWithdrawalQueueUntil(uint256 _resumeSince) internal {
+        require(_resumeSince > block.timestamp, "UNTIL_TIMESTAMP_MUST_BE_IN_FUTURE");
+        uint256 duration = _resumeSince - block.timestamp;
+
+        // TODO mainnet: recheck this requirement and maybe uncomment
+        // require(duration % SECONDS_PER_BLOCK == 0, "UNTIL_TIMESTAMP_MUST_BE_FACTOR_OF_12");
+
+        IWithdrawalQueue queue = IWithdrawalQueue(_locator.withdrawalQueue());
+
+        // Need to resume first, otherwise cannot pause
```

Comment:
?

Reply:
PausableUntil.sol:
```solidity
function _pause(uint256 _duration) internal whenResumed {
}
```

Reply:
Btw I think its better to refactor PausableUntil. Let it assume timestamp to resume since. Because otherwise it requires   additional on-chain code / contract deployed (not only here, but in general situation) to pause until specific time. And this seems to be the more frequent case, I guess.
What do you think?